### PR TITLE
LDGSTS, LDGDEPBAR and DEPBAR Implementations

### LINKED ISSUE #211: LDGSTS Support for Ampere
Including LDGSTS, LDGDEPBAR, and DEPBAR instructions. And to add the immediate field into the tracer tool. And architectural changes will be updated in the gpgpu-sim repo.

### PR DESCRIPTION
**Pull request by Connie120**
This PR includes the LDGSTS, LDGDEPBAR and DEPBAR implementations that's specifically for the NVIDIA Ampere architecture.

### CODE REVIEW

**Comment by FJShen** [comment]
Please add comments to your code citing the specification of, or an explanation for, what these instructions are supposed to do.

**Comment by JRPan** [comment]
It's probably better to assign the same guy the review the frontend and the backend

File: src/gpgpu-sim/shader.h [hunk 1]
```diff
@@ -72,6 +72,9 @@
 
 #define WRITE_MASK_SIZE 8
 
+#define SID 41
+#define WID 2
```
Comment (FJShen) [inline src/gpgpu-sim/shader.h]: Get rid of all ad-hoc personal debugging code using these numbers

File: src/gpgpu-sim/shader.h [hunk 2]
```diff
@@ -927,11 +966,13 @@ class opndcoll_rfu_t {  // operand collector based register file unit
...
     bool is_free() { return m_free; }
 
+    warp_inst_t *m_warp;
+
    private:
     bool m_free;
     unsigned m_cuid;  // collector unit hw id
     unsigned m_warp_id;
-    warp_inst_t *m_warp;
+    // warp_inst_t *m_warp;
```
Comment (FJShen) [inline src/gpgpu-sim/shader.h]: Is it necessary to turn m_warp from private to public?

File: src/gpgpu-sim/shader.cc [hunk 3]
```diff
@@ -3279,7 +3412,7 @@ void shader_core_ctx::display_pipeline(FILE *fout, int print_mem,
     if (!m_warp[i]->ibuffer_empty()) m_warp[i]->print_ibuffer(fout);
   }
   fprintf(fout, "\n");
-  display_simt_state(fout, mask);
+  // display_simt_state(fout, mask);
```
Comment (FJShen) [inline src/gpgpu-sim/shader.cc]: Why comment out this line?

File: src/gpgpu-sim/shader.h [hunk 4]
```diff
@@ -1314,6 +1355,13 @@ class ldst_unit : public pipelined_simd_unit {
             const memory_config *mem_config, class shader_core_stats *stats,
             unsigned sid, unsigned tpc);
 
+  // Ni: Add a structure to record the LDGSTS instructions,
+  // similar to m_pending_writes, but since LDGSTS does not have a output register
+  // to write to, so a new structure needs to be added
+  std::map<unsigned /*warp_id*/,
+           std::map<unsigned /*pc*/, 
+                  std::map<unsigned /*addr*/, unsigned /*count*/>>>
+      m_pending_ldgsts;
```
Comment (FJShen) [inline src/gpgpu-sim/shader.h:1361]: Nice documentation for the meaning of each level of key. However you can further
improve the comments to make it compatible with common documentation tools, such
that the meanings of each key can show up when you hover 
your mouse over the variable name. Try this:

```suggestion
  /**
   * A multi-level map: unsigned (warp_id) -> unsigned (pc) -> unsigned (addr) -> unsigned (count)
   */
  std::map<unsigned /*warp_id*/,
           std::map<unsigned /*pc*/, 
                  std::map<unsigned /*addr*/, unsigned /*count*/>>>
      m_pending_ldgsts;
```

**Review by FJShen — CHANGES_REQUESTED** [review]
Please address my comments

**Review by FJShen — APPROVED** [review]

**Comment by YWHyuk** [comment]
Hi @Connie120, I'm simulating Ampere architecture using your #62 features.

When I run the simulator, a deadlock occurs. Upon investigating the details, it seems that the warps' `m_waiting_ldgsts` is set to 1, and there is no place where `m_waiting_ldgsts` is reset to 0, causing the deadlock.

Also, I reviewed your PR and I'm curious where the part that increases the counter of `m_pending_ldgsts` is located.

**Comment by Connie120** [comment]
@YWHyuk Please check the unset_depbar function in shader.cc: https://github.com/accel-sim/gpgpu-sim_distribution/blob/7dc9977199f6a79f58191b18b64568b6155873ac/src/gpgpu-sim/shader.cc#L1849, which sets m_waiting_ldgsts to false. It's called at several places, and you can find it in shader.cc. 

For m_pending_ldgsts, please try to add the following line **within the outer if statement** after https://github.com/accel-sim/gpgpu-sim_distribution/blob/7dc9977199f6a79f58191b18b64568b6155873ac/src/gpgpu-sim/shader.cc#L2644
```
if (inst->m_is_ldgsts) {
      m_pending_ldgsts[warp_id][inst->pc][inst->get_addr(0)] += n_accesses;
}
```
I think I accidentally deleted this, thanks for bringing it up. Adding these lines could be a quick fix. Also, please let me know what app you are using so I can also test it on my end. In the meantime, if it works for you, feel free to create a PR. Or if it still doesn't work, let me know, I'll try to fix it.

**Comment by YWHyuk** [comment]
Thank you @Connie120  for quick response :)

I applied the method you mentioned, and with one additional minor fix, **it works now!**

I have attached the trace of my application. This application is CUTLASS's GEMM kernel using Tensor Cores of the Ampere architecture, and the size is 256x256x256 (MxNxK).

[gemm_256.zip](https://github.com/accel-sim/gpgpu-sim_distribution/files/15437685/gemm_256.zip)

I used the SM86_RTX3070 configuration, and setting `-gpgpu_n_clusters` to 1 reproduces the situation where the deadlock occurs.

The minor fix I made is an exception handling for the following situation.
```
2050 ffffffff 1 R72 LDSM.16.M88.4 1 R66 2 2 0x6000 32 32 32 432 32 32 32 432 -32 96 -32 496 -32 96 -32 432 32 32 32 432 32 32 32 432 -32 96 -32 496 -32 96 -32 0 
2060 ffffffff 1 R76 LDSM.16.M88.4 1 R66 2 2 0x7000 32 32 32 432 32 32 32 432 -32 96 -32 496 -32 96 -32 432 32 32 32 432 32 32 32 432 -32 96 -32 496 -32 96 -32 0 
2070 00000000 0 BRA 0 0 6352 
2080 ffffffff 0 LDGDEPBAR 0 0 0  <--------------------------- m_ldgdepbar_id++
2090 ffffffff 0 ISETP.NE.U32.AND 1 R255 0 0                   No ldgsts instruction between LDGDEPBARs
20a0 ffffffff 0 LDGDEPBAR 0 0 0  <--------------------------- m_ldgdepbar_id++
20b0 ffffffff 0 ISETP.NE.AND.EX 1 R255 0 0 
20c0 ffffffff 0 DEPBAR.LE 0 0 0 
20d0 ffffffff 0 BAR.SYNC.DEFER_BLOCKING 0 0 0 
```

I'll send these fixes as a PR.

**Comment by Connie120** [comment]
@YWHyuk Glad it works! Please add me as the reviewer of your PR so I can take a look, thanks!

**Pull request merged**